src/raw/mod.rs

```diff
@@ -1291,48 +1298,46 @@ impl<T, A: Allocator> RawTable<T, A> {
         &mut self,
         hashes: [u64; N],
         eq: impl FnMut(usize, &T) -> bool,
-    ) -> Option<[&'_ mut T; N]> {
+    ) -> [Option<&'_ mut T>; N] {
         unsafe {
-            let ptrs = self.get_many_mut_pointers(hashes, eq)?;
+            let ptrs = self.get_many_mut_pointers(hashes, eq);
 
-            for (i, &cur) in ptrs.iter().enumerate() {
-                if ptrs[..i].iter().any(|&prev| ptr::eq::<T>(prev, cur)) {
-                    return None;
+            for (i, cur) in ptrs.iter().enumerate() {
+                if cur.is_some() && ptrs[..i].contains(cur) {
+                    panic!("duplicate keys found");
                 }
             }
             // All bucket are distinct from all previous buckets so we're clear to return the result
             // of the lookup.
 
-            // TODO use `MaybeUninit::array_assume_init` here instead once that's stable.
-            Some(mem::transmute_copy(&ptrs))
+            ptrs.map(|ptr| ptr.map(|mut ptr| ptr.as_mut()))
         }
     }
 
     pub unsafe fn get_many_unchecked_mut<const N: usize>(
         &mut self,
         hashes: [u64; N],
         eq: impl FnMut(usize, &T) -> bool,
-    ) -> Option<[&'_ mut T; N]> {
-        let ptrs = self.get_many_mut_pointers(hashes, eq)?;
-        Some(mem::transmute_copy(&ptrs))
+    ) -> [Option<&'_ mut T>; N] {
+        let ptrs = self.get_many_mut_pointers(hashes, eq);
+        ptrs.map(|ptr| ptr.map(|mut ptr| ptr.as_mut()))
     }
 
     unsafe fn get_many_mut_pointers<const N: usize>(
         &mut self,
         hashes: [u64; N],
         mut eq: impl FnMut(usize, &T) -> bool,
-    ) -> Option<[*mut T; N]> {
+    ) -> [Option<NonNull<T>>; N] {
         // TODO use `MaybeUninit::uninit_array` here instead once that's stable.
-        let mut outs: MaybeUninit<[*mut T; N]> = MaybeUninit::uninit();
+        let mut outs: MaybeUninit<[Option<NonNull<T>>; N]> = MaybeUninit::uninit();
         let outs_ptr = outs.as_mut_ptr();
 
         for (i, &hash) in hashes.iter().enumerate() {
-            let cur = self.find(hash, |k| eq(i, k))?;
-            *(*outs_ptr).get_unchecked_mut(i) = cur.as_mut();
+            let cur = self.find(hash, |k| eq(i, k)).map(|cur| cur.as_non_null());
+            *(*outs_ptr).get_unchecked_mut(i) = cur;
```

Comment:
This is pre-existing but this line creates a `&mut uninit` which the UB-ness of that is still undecided https://github.com/rust-lang/unsafe-code-guidelines/issues/346. Let me change it to pure raw pointers instead to be on the safe side.

Reply:
Done in [`b582436` (#562)](https://github.com/rust-lang/hashbrown/pull/562/commits/b582436f06c080198f388ce2e2eb4030acc64afa)

Reply:
Couldn't you still do that with the `NonNull`?

Reply:
Of course, fixed in [`6b288355`](https://github.com/rust-lang/hashbrown/pull/562/commits/6b2883557ff39d6f63c050b495dd277718efc77b).

*Sorry for that, I'm working with two different computers and forgot to pull on the first one. My bad.*